### PR TITLE
[Discover] Refactor SetupAccess Screens

### PR DESCRIPTION
**Pull request by kimlisa**
part of https://github.com/gravitational/teleport/issues/16858

Note: Easier to review by commit, tried to contain specific changes. I think half of the changes are file deletes, and simple yaml file addition and changes.

#### Description
- The SetupAccess screen used similar logic, so refactored to reduce duplication
- Refactored existing kube and server SetupAccess screen
- Tentatively created database SetupAccess screen
- Add new fields from querying database resource: https://github.com/gravitational/teleport/pull/18041

#### TODO
- [ ] backport to v11
- [ ] backport to v10

#### Figma - DB setup access screen
https://www.figma.com/file/8DQZl1mYUPLgvKgmA6zPOM/In-Progress-Work?node-id=1299%3A69270

#### Screenshot
Looks pretty sim to figma, all other states are available in storybook under `Teleport/Discover/*`

### CODE REVIEW

File: packages/teleport/src/Discover/Shared/SetupAccess/useUserTraits.ts [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+  // The resource itself contains traits that define both
+  // dynamic (user-defined) and static (role-defined) traits.
+  let meta = props.agentMeta;
+  let staticTraits = initUserTraits();
+  switch (props.selectedResourceKind) {
+    case ResourceKind.Kubernetes:
+      const kube = (meta as KubeMeta).kube;
+      staticTraits.kubeUsers = kube.users.filter(
+        v => !dynamicTraits.kubeUsers.includes(v)
+      );
```
Comment (JanKaczmarkiewicz) [inline packages/teleport/src/Discover/Shared/SetupAccess/useUserTraits.ts:56]: Can be generalized into difference function (https://proofwiki.org/wiki/Definition:Set_Difference):
```suggestion
const except = (left: string[], right: string[]) =>
  left.filter(l => !right.includes(l));
```

File: packages/teleport/src/Discover/Shared/SetupAccess/SetupAccessWrapper.tsx [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+      break;
+  }
+
+  return (
+    <Box maxWidth="700px">
+      <Header>Set Up Access</Header>
+      <HeaderSubtitle>{headerSubtitle}</HeaderSubtitle>
+      {infoContent}
+      <>
```
Comment (JanKaczmarkiewicz) [inline packages/teleport/src/Discover/Shared/SetupAccess/SetupAccessWrapper.tsx:134]: No need for fragment. Maybe we should enable https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-no-useless-fragment.md?

File: packages/teleport/src/Discover/Server/SetupAccess/SetupAccess.tsx [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      traitDescription="users"
+      hasTraits={hasTraits}
+      onProceed={handleOnProceed}
+    >
+      <Box mb={2}>
+        OS Users
+        <SelectCreatable
+          inputValue={loginInputValue}
+          isClearable={selectedLogins.some(v => !v.isFixed)}
+          onInputChange={input => setLoginInputValue(input)}
```
Comment (JanKaczmarkiewicz) [inline packages/teleport/src/Discover/Server/SetupAccess/SetupAccess.tsx:91]:
```suggestion
          onInputChange={setLoginInputValue}
```

File: packages/teleport/src/Discover/Server/SetupAccess/SetupAccess.tsx [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    ...restOfProps
+  } = props;
+  const [loginInputValue, setLoginInputValue] = useState('');
+  const [selectedLogins, setSelectedLogins] = useState<Option[]>([]);
+
+  useEffect(() => {
+    if (props.attempt.status === 'success') {
+      setSelectedLogins(initSelectedOptions('logins'));
+    }
+  }, [props.attempt]);
```
Comment (JanKaczmarkiewicz) [inline packages/teleport/src/Discover/Server/SetupAccess/SetupAccess.tsx:52]:
```suggestion
  }, [props.attempt.status, initSelectedOptions]);
```

File: packages/teleport/src/Discover/Shared/SetupAccess/SetupAccessWrapper.tsx (hunk 2 above)
Comment (kimlisa) [reply packages/teleport/src/Discover/Shared/SetupAccess/SetupAccessWrapper.tsx:134]: it actually didn't help me catch it (b/c the rule is `more than one children` 😞) but I still added the rule anyways

File: packages/build/.eslintrc.js [hunk 5]
```diff
@@ -134,6 +140,10 @@ module.exports = {
     'react/self-closing-comp': 0,
     'react/sort-comp': 0,
     'react/jsx-wrap-multilines': 1,
+    'react/jsx-no-useless-fragment': 2,
+
+    'react-hooks/rules-of-hooks': 'warn',
+    'react-hooks/exhaustive-deps': 'warn',
```
Comment (hatched) [inline packages/build/.eslintrc.js:146]: Maybe keep this syntax consistent, so swap these for `1`'s

File: packages/teleport/src/Discover/Shared/SetupAccess/useUserTraits.ts [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+
+    default:
+      throw new Error(
+        `useUserTraits.ts:statiTraits: resource kind ${props.selectedResourceKind} is not handled`
+      );
+  }
+
+  useEffect(() => {
+    fetchUserTraits();
+  }, []);
```
Comment (hatched) [inline packages/teleport/src/Discover/Shared/SetupAccess/useUserTraits.ts:90]: If leaving the dependencies blank is intentional can you add a comment and a `//eslint ignore` label here so a future dev knows this was an intentional decision.

File: packages/teleport/src/mocks/contexts.ts [hunk 7]
```diff
@@ -67,3 +67,22 @@ export const userContext = makeUserContext({
     proxyVersion: '4.4.0-dev',
   },
 });
+
+export const baseContext = {
+  authType: 'local',
+  userName: 'llama',
+  accessCapabilities: {
+    suggestedReviewers: ['george_washington@gmail.com', 'alpha'],
```
Comment (hatched) [inline packages/teleport/src/mocks/contexts.ts:75]: Ahead of his time... 😉

**Review by hatched — APPROVED** [review]

File: packages/teleport/src/Discover/Shared/SetupAccess/useUserTraits.ts (hunk 6 above)
Comment (kimlisa) [reply packages/teleport/src/Discover/Shared/SetupAccess/useUserTraits.ts:90]: i used this technique: https://gravitational.slack.com/archives/C02DQ1C2BMW/p1667381168976679?thread_ts=1667257043.804519&cid=C02DQ1C2BMW

**Comment by kimlisa** [comment]
@JanKaczmarkiewicz i moved adding new eslint rules and fixing lint errors in this PR: https://github.com/gravitational/webapps/pull/1327

**Comment by kimlisa** [comment]
@JanKaczmarkiewicz friendly ping

**Review by JanKaczmarkiewicz — APPROVED** [review]

**Pull request merged**